### PR TITLE
#161255190 implement authentication on routes

### PR DESCRIPTION
**Pull request by p8ul**
[Delivers #161255190]

### What does this PR do?
- Implement authentication on routes

### Description of Task to be completed?
- create protected route
- add 404 component
- toggle authentication links based on authorization

### How should this be manually tested?
```yarn test```

### Any background context you want to provide?
https://reacttraining.com/react-router/web/example/auth-workflow

### What are the relevant pivotal tracker stories?
[#161255190](https://www.pivotaltracker.com/story/show/161255190)

### Screenshots (if appropriate)
screen shot 2018-12-09 at 10 40 35

### Checklist:
- [x] My code follows the style guidelines of this project
- [x] My teammates have reviewed my PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My PR has one commit.

### CODE REVIEW

**Review by mmosoroohh — APPROVED** [review]
This is great work on implementing the private route...

File: src/components/Menu/AuthMenuItem.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            <div className="item">
+              <a href="/signup">
+                <span className="ui primary button theme-button-color">Sign Up</span>
+              </a>
+            </div>
+          </React.Fragment>
+        )
+        : (
+          <div className="item">
+            <a href="#logout" id="logout" onClick={() => removeToken()}>
```
Comment (Rodgers-M) [inline src/components/Menu/AuthMenuItem.js:30]: Just removing the token on logout is not enough, you need to send a request to the API to invalidate the token. I also think its good practice to redirect a user after logging out, either to the home/landing page.

File: src/components/Menu/AuthMenuItem.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+import { removeToken } from '../../utils';
```
Comment (Rodgers-M) [inline src/components/Menu/AuthMenuItem.js]: i wouldn't really advocate doing this kind of import, its hard to trace where this function is defined even after visiting the index file in utils folder, for ease i would prefer importing it directly from the auth file
`import { removeToken } from '../../utils/auth`

File: src/components/Menu/__tests__/auth-menu.test.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    <BrowserRouter>
+      <AuthMenuItem authenticated={false} />
+    </BrowserRouter>,
+  );
+  expect(wrapper.find('#login')).toHaveLength(1);
+});
+
+it('renders AuthMenuItem authentication component without crashing', () => {
+  const wrapper = mount(
+    <BrowserRouter>
```
Comment (Rodgers-M) [inline src/components/Menu/__tests__/auth-menu.test.js]: you can also have a look at [Browser router context](https://www.npmjs.com/package/react-router-enzyme-context) as it solves this problem too. though it's optional, if rapping the component with browser router is working fine.

File: src/containers/Login/index.js [hunk 4]
```diff
@@ -1,8 +1,10 @@
 import React, { Component } from 'react';
-import { bindActionCreators } from 'redux';
 import { connect } from 'react-redux';
-import { LogInRequest } from '../../redux/actions/login';
 import LoginForm from '../../components/Login';
+import { setToken, removeToken } from '../../utils';
+import {
+  loginError, loginSuccess, removeMessage, LogInRequest,
```
Comment (Rodgers-M) [inline src/containers/Login/index.js:7]: you can add these to `mapStateToProps` so you don't have to call `dispatch` explicitly

Comment (p8ul) [reply src/containers/Login/index.js:7]: I'm afraid mapStateToProps won't dispatch actions. Did you mean mapDispatchtoProps?

File: src/components/Menu/AuthMenuItem.js (hunk 1 above)
Comment (p8ul) [reply src/components/Menu/AuthMenuItem.js:30]: Thank you for the feedback. I have noticed that in the backend, we don't have a logout API to blacklist/invalidate the token.

File: src/components/Menu/__tests__/auth-menu.test.js (hunk 3 above)
Comment (p8ul) [reply src/components/Menu/__tests__/auth-menu.test.js]: This is really helpful. Thank you

File: src/containers/Login/index.js (hunk 4 above)
Comment (Rodgers-M) [reply src/containers/Login/index.js:7]: yeah sure i meant  `mapDispatchtoProps`

**Pull request merged**